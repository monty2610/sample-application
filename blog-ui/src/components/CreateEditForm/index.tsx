import React, { useState, FC } from "react";
import { Form, FormGroup, Label, Input, Button } from "reactstrap";
import { Link } from "react-router-dom";

type Blog = {
  title: string;
  content: string;
};
type FormProps = {
  initialValues?: Blog;
  onSubmitHandler: (updatedContent: Blog) => void;
};

const CreateEditForm: FC<FormProps> = ({ initialValues, onSubmitHandler }) => {
  const [title, setTitle] = useState(
    (initialValues && initialValues.title) || ""
  );
  const [content, setContent] = useState(
    (initialValues && initialValues.content) || ""
  );

  const onInputChange = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    const element = event.target as HTMLInputElement;
    if (element.name === "title") {
      setTitle(element.value);
    } else if (element.name === "content") {
      setContent(element.value);
    }
  };

  const onSubmit = (event: React.FormEvent) => {
    event.preventDefault();
    if (title && content) {
      onSubmitHandler({ title, content });
    }
  };

  return (
    <Form onSubmit={onSubmit}>
      <FormGroup>
        <Label for="title">Title</Label>
        <Input
          name="title"
          id="title"
          placeholder="Write a Title"
          value={title}
          onChange={onInputChange}
        />
      </FormGroup>
      <FormGroup>
        <Label for="content">Content</Label>
        <Input
          type="textarea"
          name="content"
          id="content"
          placeholder="Content goes here..."
          value={content}
          onChange={onInputChange}
        />
      </FormGroup>
      <div className="align-right">
        <div>
          <Button to={"/"} tag={Link} className="mr" size="sm">
            Back
          </Button>

          <Button type="submit" size="sm" color="primary">
            Submit
          </Button>
        </div>
      </div>
    </Form>
  );
};

export default CreateEditForm;
