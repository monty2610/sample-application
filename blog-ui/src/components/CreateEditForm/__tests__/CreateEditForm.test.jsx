import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import CreateEditForm from "../index";

const getContainer = () => {
  return render(
    <BrowserRouter>
      <CreateEditForm />
    </BrowserRouter>
  );
};

describe("CreateEditForm component", () => {
  let container;
  let onSubmitHandler;
  beforeEach(() => {
    onSubmitHandler = jest.fn();
    container = render(
      <BrowserRouter>
        <CreateEditForm onSubmitHandler={onSubmitHandler} />
      </BrowserRouter>
    );
  });

  it("should match snapshot correctly", () => {
    const { asFragment } = container;

    expect(asFragment()).toMatchSnapshot();
  });
});

describe("CreateEditForm component with initialValues prop", () => {
  let container;
  let onSubmitHandler;

  beforeEach(() => {
    onSubmitHandler = jest.fn();
    container = render(
      <BrowserRouter>
        <CreateEditForm
          initialValues={{ title: "test post", content: "test content" }}
          onSubmitHandler={onSubmitHandler}
        />
      </BrowserRouter>
    );
  });

  it("should match snapshot correctly", () => {
    const { asFragment } = container;

    expect(asFragment()).toMatchSnapshot();
  });

  it("initial values should be populated correctly", () => {
    const { getByPlaceholderText } = container;
    expect(getByPlaceholderText("Write a Title")).toHaveValue("test post");
    expect(getByPlaceholderText("Content goes here...")).toHaveValue(
      "test content"
    );
  });

  it("onSubmit should call the onSubmitHandler prop with updated content", () => {
    const { getByPlaceholderText, getByRole } = container;
    fireEvent.change(getByPlaceholderText("Write a Title"), {
      target: {
        value: "new post",
      },
    });

    fireEvent.change(getByPlaceholderText("Content goes here..."), {
      target: {
        value: "new content",
      },
    });

    fireEvent.click(
      getByRole("button", {
        name: "Submit",
      })
    );

    expect(onSubmitHandler).toHaveBeenCalledWith({
      title: "new post",
      content: "new content",
    });
  });
});
