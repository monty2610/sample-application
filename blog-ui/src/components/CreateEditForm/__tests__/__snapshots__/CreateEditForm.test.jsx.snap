// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEditForm component should match snapshot correctly 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="form-group"
    >
      <label
        class=""
        for="title"
      >
        Title
      </label>
      <input
        class="form-control"
        id="title"
        name="title"
        placeholder="Write a Title"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class=""
        for="content"
      >
        Content
      </label>
      <textarea
        class="form-control"
        id="content"
        name="content"
        placeholder="Content goes here..."
      />
    </div>
    <div
      class="align-right"
    >
      <div>
        <a
          class="mr btn btn-secondary btn-sm"
          href="/"
        >
          Back
        </a>
        <button
          class="btn btn-primary btn-sm"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`CreateEditForm component with initialValues prop should match snapshot correctly 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="form-group"
    >
      <label
        class=""
        for="title"
      >
        Title
      </label>
      <input
        class="form-control"
        id="title"
        name="title"
        placeholder="Write a Title"
        type="text"
        value="test post"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class=""
        for="content"
      >
        Content
      </label>
      <textarea
        class="form-control"
        id="content"
        name="content"
        placeholder="Content goes here..."
      >
        test content
      </textarea>
    </div>
    <div
      class="align-right"
    >
      <div>
        <a
          class="mr btn btn-secondary btn-sm"
          href="/"
        >
          Back
        </a>
        <button
          class="btn btn-primary btn-sm"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
